File: src/commonMain/kotlin/mahjongutils/shanten/Shanten.kt
package mahjongutils.shanten

import mahjongutils.models.Furo
import mahjongutils.models.Tile
import mahjongutils.models.countAsCodeArray
import mahjongutils.models.hand.CommonHandPattern
import mahjongutils.models.hand.Hand
import kotlin.math.min


private fun <S : CommonShanten> mergeIntoWithoutGot(
    targetShantenNum: Int,
    advance: MutableSet<Tile>,
    goodShapeAdvance: MutableSet<Tile>,
    patterns: MutableCollection<CommonHandPattern>,
    result: ShantenResult<S, *>
) {
    if (result.shantenInfo.shantenNum == targetShantenNum) {
        val shantenInfo = result.shantenInfo.asWithoutGot
        advance += shantenInfo.advance
        patterns += result.hand.patterns
        if (shantenInfo.goodShapeAdvance != null) {
            goodShapeAdvance += shantenInfo.goodShapeAdvance
        }
    }
}

private fun <S : CommonShanten> mergeIntoWithGot(
    targetShantenNum: Int,
    discardToAdvance: MutableMap<Tile, ShantenWithoutGot>,
    patterns: MutableCollection<CommonHandPattern>,
    result: ShantenResult<S, *>,
    bestShantenOnly: Boolean
) {
    val shantenInfo = result.shantenInfo.asWithGot
    for ((discard, shantenAfterDiscard) in shantenInfo.discardToAdvance) {
        if (!bestShantenOnly || shantenAfterDiscard.shantenNum == targetShantenNum) {
            val oldShantenAfterDiscard = discardToAdvance[discard]
            if (oldShantenAfterDiscard == null || oldShantenAfterDiscard.shantenNum > shantenAfterDiscard.shantenNum) {
                discardToAdvance[discard] = shantenAfterDiscard
            } else if (oldShantenAfterDiscard.shantenNum == shantenAfterDiscard.shantenNum) {
                discardToAdvance[discard] = oldShantenAfterDiscard.copy(
                    advance = oldShantenAfterDiscard.advance + shantenAfterDiscard.advance,
                    goodShapeAdvance = shantenAfterDiscard.goodShapeAdvance?.let {
                        oldShantenAfterDiscard.goodShapeAdvance?.plus(it)
                    }
                )
            }
        }
    }

    if (targetShantenNum == shantenInfo.shantenNum) {
        patterns += result.hand.patterns
    }
}

/**
 * 向听分析
 * @param tiles 门前的牌
 * @param furo 副露（对向听分析本身无用，但若需要将结果用于和了分析则需要传入）
 * @param bestShantenOnly 仅计算最优向听数的打法（不计算退向打法）
 * @return 向听分析结果
 */
fun shanten(
    tiles: List<Tile>, furo: List<Furo> = emptyList(),
    bestShantenOnly: Boolean = false
): UnionShantenResult = shanten(tiles, furo, true, true, bestShantenOnly, true, true)

/**
 * 向听分析
 * @param tiles 门前的牌
 * @param furo 副露（对向听分析本身无用，但若需要将结果用于和了分析则需要传入）
 * @param calcAdvanceNum 是否计算进张数
 * @param calcGoodShapeAdvance 是否计算好型进张
 * @param bestShantenOnly 仅计算最优向听数的打法（不计算退向打法）
 * @param allowAnkan 是否允许暗杠
 * @param calcImprovement 是否计算改良
 * @return 向听分析结果
 */
internal fun shanten(
    tiles: List<Tile>, furo: List<Furo> = emptyList(),
    calcAdvanceNum: Boolean = true,
    calcGoodShapeAdvance: Boolean = true,
    bestShantenOnly: Boolean = false,
    allowAnkan: Boolean = true,
    calcImprovement: Boolean = true,
): UnionShantenResult {
    val tiles = ensureLegalTiles(tiles)

    val withGot = tiles.size % 3 == 2
    val k = tiles.size / 3

    if (k != 4) {
        val regular = regularShanten(
            tiles, furo,
            calcAdvanceNum = calcAdvanceNum,
            calcGoodShapeAdvance = calcGoodShapeAdvance,
            bestShantenOnly = bestShantenOnly,
            allowAnkan = allowAnkan,
            calcImprovement = calcImprovement
        )
        return UnionShantenResult(
            hand = regular.hand, shantenInfo = regular.shantenInfo,
            regular = regular
        )
    }

    val regular = regularShanten(
        tiles, furo,
        calcAdvanceNum = false,
        calcGoodShapeAdvance = calcGoodShapeAdvance,
        bestShantenOnly = bestShantenOnly,
        allowAnkan = allowAnkan,
        calcImprovement = calcImprovement
    )
    val chitoi = chitoiShanten(tiles, calcAdvanceNum = false, bestShantenOnly = bestShantenOnly)
    val kokushi = kokushiShanten(tiles, calcAdvanceNum = false, bestShantenOnly = bestShantenOnly)

    val shantenNum = min(
        min(
            regular.shantenInfo.shantenNum,
            chitoi.shantenInfo.shantenNum
        ),
        kokushi.shantenInfo.shantenNum
    )
    val patterns = ArrayList<CommonHandPattern>()

    var shantenInfo = if (!withGot) {
        val advance = HashSet<Tile>()
        val goodShapeAdvance = HashSet<Tile>()

        mergeIntoWithoutGot(shantenNum, advance, goodShapeAdvance, patterns, regular)
        mergeIntoWithoutGot(shantenNum, advance, goodShapeAdvance, patterns, chitoi)
        mergeIntoWithoutGot(shantenNum, advance, goodShapeAdvance, patterns, kokushi)

        ShantenWithoutGot(
            shantenNum, advance,
            goodShapeAdvance = if (shantenNum == 1) goodShapeAdvance else null,
            improvement = (regular.shantenInfo.asWithoutGot).improvement
                ?: (if (shantenNum == 0) emptyMap() else null),
            goodShapeImprovement = (regular.shantenInfo.asWithoutGot).goodShapeImprovement
                ?: (if (shantenNum == 0) emptyMap() else null)
        )
    } else {
        val discardToAdvance = HashMap<Tile, ShantenWithoutGot>()

        mergeIntoWithGot(shantenNum, discardToAdvance, patterns, regular, bestShantenOnly)
        mergeIntoWithGot(shantenNum, discardToAdvance, patterns, chitoi, bestShantenOnly)
        mergeIntoWithGot(shantenNum, discardToAdvance, patterns, kokushi, bestShantenOnly)

        ShantenWithGot(
            shantenNum, discardToAdvance,
            ankanToAdvance = (regular.shantenInfo.asWithGot).ankanToAdvance
        )
    }

    if (calcAdvanceNum) {
        val tilesCount = (tiles + furo.flatMap { it.asMentsu().tiles }).countAsCodeArray()
        shantenInfo = shantenInfo.fillNum(tilesCount)
    }

    val hand = Hand(tiles = tiles, furo = furo, patterns = patterns)
    return UnionShantenResult(
        hand = hand, shantenInfo = shantenInfo,
        regular = regular, chitoi = chitoi, kokushi = kokushi
    )
}

File: python-lib/mahjong_utils/shanten/shanten.py
from typing import Optional, Sequence

from mahjong_utils.bridge import bridge_mahjongutils
from mahjong_utils.models.furo import Furo
from mahjong_utils.models.tile import Tile
from mahjong_utils.shanten.models import ShantenResult, UnionShantenResult, KokushiShantenResult, ChitoiShantenResult, \
    RegularShantenResult, FuroChanceShantenResult


def regular_shanten(
        tiles: Sequence[Tile],
        furo: Optional[Sequence[Furo]] = None,
        best_shanten_only: bool = False
) -> RegularShantenResult:
    """
    标准形向听分析

    :param tiles: 门前的牌
    :param furo: 副露（对向听分析本身无用，但若需要将结果用于和了分析则需要传入）
    :param best_shanten_only: 仅计算最优向听数的打法（不计算退向打法）
    :return 向听分析结果
    """
    result = bridge_mahjongutils.call("regularShanten", {
        "tiles": [str(t) for t in tiles],
        "furo": [fr.__encode__() for fr in furo] if furo is not None else [],
        "bestShantenOnly": best_shanten_only
    })

    return RegularShantenResult.__decode__(result)


def chitoi_shanten(
        tiles: Sequence[Tile],
        best_shanten_only: bool = False,
) -> ChitoiShantenResult:
    """
    七对子向听分析

    :param tiles: 门前的牌
    :param best_shanten_only: 仅计算最优向听数的打法（不计算退向打法）
    :return 向听分析结果
    """
    result = bridge_mahjongutils.call("chitoiShanten", {
        "tiles": [str(t) for t in tiles],
        "bestShantenOnly": best_shanten_only,
    })

    return ChitoiShantenResult.__decode__(result)


def kokushi_shanten(
        tiles: Sequence[Tile],
        best_shanten_only: bool = False,
) -> KokushiShantenResult:
    """
    国士无双向听分析

    :param tiles: 门前的牌
    :param best_shanten_only: 仅计算最优向听数的打法（不计算退向打法）
    :return 向听分析结果
    """
    result = bridge_mahjongutils.call("kokushiShanten", {
        "tiles": [str(t) for t in tiles],
        "bestShantenOnly": best_shanten_only,
    })

    return KokushiShantenResult.__decode__(result)


def shanten(
        tiles: Sequence[Tile],
        furo: Optional[Sequence[Furo]] = None,
        best_shanten_only: bool = False
) -> UnionShantenResult:
    """
    向听分析

    :param tiles: 门前的牌
    :param furo: 副露（对向听分析本身无用，但若需要将结果用于和了分析则需要传入）
    :param best_shanten_only: 仅计算最优向听数的打法（不计算退向打法）
    :return 向听分析结果
    """
    result = bridge_mahjongutils.call("shanten", {
        "tiles": [str(t) for t in tiles],
        "furo": [fr.__encode__() for fr in furo] if furo is not None else [],
        "bestShantenOnly": best_shanten_only
    })

    return UnionShantenResult.__decode__(result)


def furo_chance_shanten(
        tiles: Sequence[Tile],
        chance_tile: Tile,
        allow_chi: bool = True,
        best_shanten_only: bool = False,
        allow_kuikae: bool = False
):
    """
    副露判断分析

    :param tiles: 门前的牌
    :param chance_tile: 副露机会牌（能够吃、碰的牌）
    :param allow_chi: 是否允许吃
    :param best_shanten_only: 仅计算最优向听数的打法（不计算退向打法）
    :param allow_kuikae: 是否允许食替
    :return 向听分析结果
    """
    result = bridge_mahjongutils.call("furoChanceShanten", {
        "tiles": [str(t) for t in tiles],
        "chanceTile": chance_tile.__encode__(),
        "allowChi": allow_chi,
        "bestShantenOnly": best_shanten_only,
        "allowKuikae": allow_kuikae
    })

    return FuroChanceShantenResult.__decode__(result)


__all__ = ("regular_shanten",
           "chitoi_shanten",
           "kokushi_shanten",
           "furo_chance_shanten",
           "shanten",
           "ShantenResult",)
